sdk/core/core-client-paging-rest/src/paginate.ts

```diff
@@ -45,13 +56,47 @@ export interface PaginateOptions {
 export function paginateResponse<TElement>(
   client: Client,
   initialResponse: HttpResponse,
-  options: PaginateOptions = {}
+  options?: PaginateOptions
+): PagedAsyncIterableIterator<TElement>;
+/**
+ * Helper to iterate pageable responses
+ * @param client - Client to use for sending the request to get additional pages
+ * @param initialResponse - The initial response
+ * @param getPage - Function to define a custom pagination strategy, implementing how to get nextLink and current page
+ */
+export function paginateResponse<TElement>(
+  client: Client,
+  initialResponse: HttpResponse,
+  getPage: GetPage<string, TElement[]>
+): PagedAsyncIterableIterator<TElement>;
+export function paginateResponse<TElement>(
+  client: Client,
+  initialResponse: HttpResponse,
+  optionsOrGetPage: PaginateOptions | GetPage<string, TElement[]> | undefined
 ): PagedAsyncIterableIterator<TElement> {
+  let options: PaginateOptions = {};
+  let getPage: GetPage<string, TElement[]> | undefined;
+  let itemName: string | undefined;
+  let nextLinkName: string | undefined;
+
+  if (typeof optionsOrGetPage === "function") {
+    getPage = optionsOrGetPage;
+  } else {
+    options = optionsOrGetPage || {};
+    const pagingInfo = getPaginationProperties(initialResponse, options);
+    itemName = pagingInfo.itemName;
+    nextLinkName = pagingInfo.nextLinkName;
+  }
+
   let firstRun = true;
-  const { itemName, nextLinkName } = getPaginationProperties(initialResponse, options);
   const pagedResult: PagedResult<TElement[]> = {
     firstPageLink: "",
     async getPage(pageLink: string) {
+      // If there is a custom getPage defined, use it insetad of the default
+      if (getPage) {
+        return getPage(pageLink);
```

Comment:
if you will never pass a max page size, then we do not have to expose that parameter in the surface